core/base/temporary_conversion.hpp

```diff
@@ -0,0 +1,227 @@
+                            ConversionCandidates...>(obj);
+    }
+
+    template <typename TargetType, typename MaybeConstLinOp,
+              typename FirstCandidate, typename... TrailingCandidates>
+    static std::unique_ptr<TargetType, std::function<void(TargetType *)>>
+    convert_impl(MaybeConstLinOp *obj)
+    {
+        using candidate_type =
+            typename std::conditional<std::is_const<MaybeConstLinOp>::value,
+                                      const FirstCandidate,
+                                      FirstCandidate>::type;
+        candidate_type *cast_obj{};
+        if ((cast_obj = dynamic_cast<candidate_type *>(obj))) {
+            auto converted = TargetType::create(obj->get_executor());
+            cast_obj->convert_to(converted.get());
+            static_assert(
+                std::is_base_of<ConvertibleTo<std::remove_cv_t<TargetType>>,
+                                FirstCandidate>::value,
+                "ConvertibleTo not implemented");
+            static_assert(std::is_base_of<ConvertibleTo<FirstCandidate>,
+                                          TargetType>::value,
+                          "ConvertibleTo not symmetric");
+            return {converted.release(),
+                    convert_back_deleter<TargetType, candidate_type>{cast_obj}};
+        } else {
+            return conversion_helper<TrailingCandidates...>::template convert<
+                TargetType>(obj);
+        }
+    }
+};
+
+template <>
+struct conversion_helper<> {
+    template <typename T, typename MaybeConstLinOp>
+    static std::unique_ptr<T, std::function<void(T *)>> convert(
+        MaybeConstLinOp *obj, bool initialize)
+    {
+        throw NotSupported(__FILE__, __LINE__,
+                           std::string{"gko::make_temporary_conversion<"} +
+                               name_demangling::get_type_name(typeid(T)) + ">",
+                           name_demangling::get_type_name(typeid(*obj)));
+        return {nullptr, null_deleter<T>{}};
+    }
+};
+
+
+/**
+ * A temporary_conversion is a special smart pointer-like object that is
+ * designed to hold an object temporarily converted to another format.
+ *
+ * After the temporary_conversion goes out of scope, the stored object will
+ * be converted back to its original format. This class is optimized to
+ * avoid copies if the object is already in the correct format, in which
+ * case it will just hold a reference to that object, without performing the
+ * conversion.
+ *
+ * @tparam T  the type of object held in the temporary_conversion
+ */
+template <typename T>
+class temporary_conversion {
+public:
+    using value_type = T;
+    using pointer = T *;
+    using lin_op_type = typename std::conditional<std::is_const<T>::value,
+                                                  const LinOp, LinOp>::type;
+
+    template <typename... Formats>
+    static temporary_conversion create(lin_op_type *ptr)
```

Comment:
Please add a small sentence what the `Formats...` types represent.
Also, I think renaming it to `FormatCandidates`, `ConversionCandidates` or similar, like you do in `conversion_helper` would be a lot better, otherwise, one might assume that you want to convert **TO** the Format (since most of the time, you only specify one type, which is `next_precision<[...]>`).